The project aims to explore the unemployment reasons for two Australian organizations DETE and TAFE through employee exit surveys. We will try to investigate the reasons that employees are resigning and compare these with the long-term employees

In [48]:
import pandas as pd
import numpy as np

#reading the data into pandas for initial exploration
dete_survey = pd.read_csv('dete_survey.csv')
tafe_survey = pd.read_csv('tafe_survey.csv')

#examining the data for dete
dete_survey.info()

dete_survey.head(10)

dete_survey['SeparationType'].value_counts()







<class 'pandas.core.frame.DataFrame'>
RangeIndex: 822 entries, 0 to 821
Data columns (total 56 columns):
ID                                     822 non-null int64
SeparationType                         822 non-null object
Cease Date                             822 non-null object
DETE Start Date                        822 non-null object
Role Start Date                        822 non-null object
Position                               817 non-null object
Classification                         455 non-null object
Region                                 822 non-null object
Business Unit                          126 non-null object
Employment Status                      817 non-null object
Career move to public sector           822 non-null bool
Career move to private sector          822 non-null bool
Interpersonal conflicts                822 non-null bool
Job dissatisfaction                    822 non-null bool
Dissatisfaction with the department    822 non-null bool
Physical work environ

Age Retirement                          285
Resignation-Other reasons               150
Resignation-Other employer               91
Resignation-Move overseas/interstate     70
Voluntary Early Retirement (VER)         67
Ill Health Retirement                    61
Other                                    49
Contract Expired                         34
Termination                              15
Name: SeparationType, dtype: int64

In [49]:
#examining data for tafe survey
tafe_survey.info()

tafe_survey.head(15)

tafe_survey['Reason for ceasing employment'].value_counts()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 702 entries, 0 to 701
Data columns (total 72 columns):
Record ID                                                                                                                                                        702 non-null float64
Institute                                                                                                                                                        702 non-null object
WorkArea                                                                                                                                                         702 non-null object
CESSATION YEAR                                                                                                                                                   695 non-null float64
Reason for ceasing employment                                                                                                                                    701 non-

Resignation                 340
Contract Expired            127
Retrenchment/ Redundancy    104
Retirement                   82
Transfer                     25
Termination                  23
Name: Reason for ceasing employment, dtype: int64

For DETE:
We can make the following observations based on the work above:
 The dete_survey dataframe contains 'Not Stated' values that indicate values are missing, but they aren't represented as NaN.
 Both the dete_survey and tafe_survey contain many columns that we don't need to complete our analysis.
Each dataframe contains many of the same columns, but the column names are different.
 There are multiple columns/answers that indicate an employee resigned because they were dissatisfied.
 The last 5 columns have alot of null values.

For TAFE: 
15 rows and 72 columns
ID had the string ex17 at end of all entries which can be dropped
Contributing factors columns have alot of missing values

Resignation                 340 
Contract Expired            127
Retrenchment/ Redundancy    104
Retirement                   82
Transfer                     25
Termination                  23

Age Retirement                          285
Resignation-Other reasons               150
Resignation-Other employer               91
Resignation-Move overseas/interstate     70
Voluntary Early Retirement (VER)         67
Ill Health Retirement                    61
Other                                    49
Contract Expired                         34
Termination                              15

In [50]:
#setting not stated values as NaN in dete survey
dete_survey = pd.read_csv('dete_survey.csv', na_values = 'Not Stated')

In [51]:
#dropping unwanted columns from dete

dete_survey_updated = dete_survey.drop(dete_survey.columns[28:49], axis=1)
tafe_survey_updated = tafe_survey.drop(tafe_survey.columns[17:66], axis=1)

columns that do not contribute to our analysis of reasons for employee exit were removed.

In [52]:
#renaming columns in dete
dete_survey_updated.columns = dete_survey_updated.columns.str.lower().str.replace(' ', '_').str.strip()
print(dete_survey_updated.head(5))

#updating column names in tafe
tafe_survey_updated = tafe_survey.rename(columns = {'CESSATION YEAR':'cease_date', 'Record ID':'id', 'Reason for ceasing employment':'separationtype', 'Gender. What is your Gender?':'gender','Employment Type. Employment Type':'employment_status', 'Classification. Classification': 'position','LengthofServiceOverall. Overall Length of Service at Institute (in years)': 'institute_service', 'LengthofServiceCurrent. Length of Service at current workplace (in years)': 'role_service'})

   id                    separationtype cease_date  dete_start_date  \
0   1             Ill Health Retirement    08/2012           1984.0   
1   2  Voluntary Early Retirement (VER)    08/2012              NaN   
2   3  Voluntary Early Retirement (VER)    05/2012           2011.0   
3   4         Resignation-Other reasons    05/2012           2005.0   
4   5                    Age Retirement    05/2012           1970.0   

   role_start_date                                      position  \
0           2004.0                                Public Servant   
1              NaN                                Public Servant   
2           2011.0                               Schools Officer   
3           2006.0                                       Teacher   
4           1989.0  Head of Curriculum/Head of Special Education   

  classification              region                      business_unit  \
0        A01-A04      Central Office  Corporate Strategy and Peformance   
1        AO5-A

In [53]:
tafe_survey_updated.head()

id                              Institute  \
0  6.341330e+17  Southern Queensland Institute of TAFE   
1  6.341337e+17            Mount Isa Institute of TAFE   
2  6.341388e+17            Mount Isa Institute of TAFE   
3  6.341399e+17            Mount Isa Institute of TAFE   
4  6.341466e+17  Southern Queensland Institute of TAFE   

                   WorkArea  cease_date    separationtype  \
0  Non-Delivery (corporate)      2010.0  Contract Expired   
1  Non-Delivery (corporate)      2010.0        Retirement   
2       Delivery (teaching)      2010.0        Retirement   
3  Non-Delivery (corporate)      2010.0       Resignation   
4       Delivery (teaching)      2010.0       Resignation   

  Contributing Factors. Career Move - Public Sector   \
0                                                NaN   
1                                                  -   
2                                                  -   
3                                                  -   
4                                                  -   

  Contributing Factors. Career Move - Private Sector   \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                       Career Move - Private Sector    

  Contributing Factors. Career Move - Self-employment  \
0                                                NaN    
1                                                  -    
2                                                  -    
3                                                  -    
4                                                  -    

  Contributing Factors. Ill Health Contributing Factors. Maternity/Family  \
0                              NaN                                    NaN   
1                                -                                      -   
2                                -                                      -   
3                                -                                      -   
4                                -                                      -   

      ...       \
0     ...        
1     ...        
2     ...        
3     ...        
4     ...        

  Workplace. Topic:Does your workplace promote a work culture free from all forms of unlawful discrimination?  \
0                                                Yes                                                            
1                                                Yes                                                            
2                                                Yes                                                            
3                                                Yes                                                            
4                                                Yes                                                            

  Workplace. Topic:Does your workplace promote and practice the principles of employment equity?  \
0                                                Yes                                               
1                                                Yes                                               
2                                                Yes                                               
3                                                Yes                                               
4                                                Yes                                               

  Workplace. Topic:Does your workplace value the diversity of its employees?  \
0                                                Yes                           
1                                                Yes                           
2                                                Yes                           
3                                                Yes                           
4                     

In [54]:
dete_survey_updated['separationtype'].value_counts()


Age Retirement                          285
Resignation-Other reasons               150
Resignation-Other employer               91
Resignation-Move overseas/interstate     70
Voluntary Early Retirement (VER)         67
Ill Health Retirement                    61
Other                                    49
Contract Expired                         34
Termination                              15
Name: separationtype, dtype: int64

In [55]:
tafe_survey_updated['separationtype'].value_counts()

Resignation                 340
Contract Expired            127
Retrenchment/ Redundancy    104
Retirement                   82
Transfer                     25
Termination                  23
Name: separationtype, dtype: int64

In [56]:
#selecting only rows with separation type resignation for dete 

pattern = r"Resignation"
resignation_in = dete_survey_updated['separationtype'].str.contains(pattern, na = False)
dete_resignations = dete_survey_updated[resignation_in].copy()

#removing characters after Resignation
dete_resignations['separationtype'] = dete_resignations['separationtype'].str.split('-').str.get(0)
dete_resignations


id separationtype cease_date  dete_start_date  role_start_date  \
3      4    Resignation    05/2012           2005.0           2006.0   
5      6    Resignation    05/2012           1994.0           1997.0   
8      9    Resignation    07/2012           2009.0           2009.0   
9     10    Resignation       2012           1997.0           2008.0   
11    12    Resignation       2012           2009.0           2009.0   
12    13    Resignation       2012           1998.0           1998.0   
14    15    Resignation       2012           2007.0           2010.0   
16    17    Resignation       2012              NaN              NaN   
20    21    Resignation       2012           1982.0           1982.0   
21    22    Resignation       2012           1980.0           2009.0   
22    23    Resignation       2012           1997.0           1998.0   
23    24    Resignation       2012           1973.0           2012.0   
25    26    Resignation       2012           1995.0           2002.0   
27    28    Resignation       2012           2005.0           2011.0   
33    34    Resignation       2012           2003.0           2003.0   
34    35    Resignation       2012           2006.0           2009.0   
37    38    Resignation       2012           2011.0           2011.0   
39    40    Resignation       2012              NaN              NaN   
40    41    Resignation       2012           1977.0           1980.0   
41    42    Resignation       2012           1974.0           1994.0   
42    43    Resignation       2012           2011.0           2011.0   
43    44    Resignation       2012           1976.0           1976.0   
48    49    Resignation       2012           2009.0           2010.0   
50    51    Resignation       2012           2009.0           2010.0   
51    52    Resignation       2012           1993.0           1993.0   
55    56    Resignation       2012           2008.0           2008.0   
57    58    Resignation       2012           2003.0           2012.0   
61    62    Resignation       2012           2011.0           2011.0   
69    70    Resignation       2012           2006.0              NaN   
71    72    Resignation       2012           2011.0           2011.0   
..   ...            ...        ...              ...              ...   
747  749    Resignation    01/2014           2008.0           2008.0   
751  753    Resignation    12/2013           2005.0           2005.0   
752  754    Resignation    11/2013           1998.0              NaN   
753  755    Resignation    11/2013           2004.0           2011.0   
755  757    Resignation    12/2013           2012.0           2013.0   
762  764    Resignation    07/2006           2006.0           2006.0   
766  768    Resignation    01/2014           2007.0           2007.0   
769  771    Resignation    11/2013           2008.0              NaN   
770  772    Resignation        NaN           1987.0           1987.0   
771  773    Resignation    01/2014           2002.0           2007.0   
774  776    Resignation        NaN           2005.0           2005.0   
784  786    Resignation    10/2013           2013.0           2013.0   
786  788    Resignation    01/2014           1994.0           2008.0   
788  790    Resignation        NaN           1990.0           2010.0   
789  791    Resignation    01/2014           1983.0           1999.0   
790  792    Resignation    01/2014           2008.0           2008.0   
791  793    Resignation        NaN           2007.0           2007.0   
794  796    Resignation    11/2013              NaN              NaN   
797  799    Resignation        NaN           2000.0           2013.0   
798  800    Resignation        NaN           1995.0              NaN   
802  804    Resignation    12/2013              NaN              NaN   
803  805    Resignation    01/2014           2004.0           2007.0   
804  806    Resignation    01/2014           2008.0           2013.0   
806  808    Resignation    12/2013           2

In [57]:
#selecting only rows with separation type resignation for tafe 
pattern_t = r"Resignation"
resignation_t = tafe_survey_updated['separationtype'].str.contains(pattern, na = False)
tafe_resignations = tafe_survey_updated[resignation_t].copy()
tafe_resignations

id                              Institute  \
3    6.341399e+17            Mount Isa Institute of TAFE   
4    6.341466e+17  Southern Queensland Institute of TAFE   
5    6.341475e+17  Southern Queensland Institute of TAFE   
6    6.341520e+17         Barrier Reef Institute of TAFE   
7    6.341537e+17  Southern Queensland Institute of TAFE   
8    6.341579e+17  Southern Queensland Institute of TAFE   
9    6.341588e+17         Barrier Reef Institute of TAFE   
10   6.341588e+17         Barrier Reef Institute of TAFE   
13   6.341725e+17         Barrier Reef Institute of TAFE   
14   6.341726e+17   Central Queensland Institute of TAFE   
15   6.341761e+17  Southern Queensland Institute of TAFE   
16   6.341770e+17       Brisbane North Institute of TAFE   
17   6.341771e+17       Brisbane North Institute of TAFE   
18   6.341779e+17       Brisbane North Institute of TAFE   
19   6.341820e+17      Southbank Institute of Technology   
20   6.341821e+17       Brisbane North Institute of TAFE   
21   6.341831e+17       Brisbane North Institute of TAFE   
22   6.341847e+17  Southern Queensland Institute of TAFE   
23   6.341907e+17   Central Queensland Institute of TAFE   
24   6.341907e+17   Central Queensland Institute of TAFE   
26   6.341934e+17            Mount Isa Institute of TAFE   
27   6.341935e+17       Brisbane North Institute of TAFE   
29   6.341994e+17         Barrier Reef Institute of TAFE   
32   6.342002e+17       Brisbane North Institute of TAFE   
36   6.342062e+17       Sunshine Coast Institute of TAFE   
37   6.342080e+17      Southbank Institute of Technology   
39   6.342081e+17      Southbank Institute of Technology   
40   6.342090e+17       Sunshine Coast Institute of TAFE   
41   6.342090e+17       Sunshine Coast Institute of TAFE   
42   6.342090e+17       Sunshine Coast Institute of TAFE   
..            ...                                    ...   
659  6.349985e+17  Southern Queensland Institute of TAFE   
660  6.349994e+17   Central Queensland Institute of TAFE   
661  6.350003e+17       Brisbane North Institute of TAFE   
665  6.350055e+17       Tropical North Institute of TAFE   
666  6.350055e+17       Tropical North Institute of TAFE   
669  6.350108e+17       Brisbane North Institute of TAFE   
670  6.350124e+17       Tropical North Institute of TAFE   
671  6.350127e+17       Tropical North Institute of TAFE   
675  6.350175e+17       Tropical North Institute of TAFE   
676  6.350194e+17  Southern Queensland Institute of TAFE   
677  6.350219e+17       Brisbane North Institute of TAFE   
678  6.350253e+17       Brisbane North Institute of TAFE   
679  6.350279e+17  Southern Queensland Institute of TAFE   
681  6.350314e+17   Central Queensland Institute of TAFE   
682  6.350357e+17            Mount Isa Institute of TAFE   
683  6.350374e+17   Central Queensland Institute of TAFE   
684  6.350375e+17       Brisbane North Institute of TAFE   
685  6.350402e+17       Brisbane North Institute of TAFE   
686  6.350426e+17       Brisbane North Institute of TAFE   
688  6.350479e+17       Tropical North Institute of TAFE   
689  6.350480e+17   Central Queensland Institute of TAFE   
690  6.350496e+17  Southern Queensland Institute of TAFE   
691  6.350496e+17  Southern Queensland Institute of TAFE   
693  6.350599e+17       Tropical North Institute of TAFE   
694  6.350652e+17       Sunshine Coast Institute of TAFE   
696  6.350660e+17  Southern Queensland Institute of TAFE   
697  6.350668e+17         Barrier Reef Institute of TAFE   
698  6.350677e+17  Southern Queensland Institute of TAFE   
699  6.350704e+17       Tropical North Institute of TAFE   
701  6.350730e+17       Tropical North Institute of TAFE   

                     WorkArea  cease_date separationtype  \
3    Non-Delivery (corporate)      2010.0    Resignation   
4         Delivery (teaching)      2010.0    Resignation   
5         Delivery (teaching)      2010.0    Resignation   
6    Non-Delivery (corporate)      2010.0    Resignati

In [58]:
dete_resignations['cease_date'].value_counts()

2012       126
2013        74
01/2014     22
12/2013     17
06/2013     14
09/2013     11
07/2013      9
11/2013      9
10/2013      6
08/2013      4
05/2013      2
05/2012      2
07/2006      1
07/2012      1
09/2010      1
2010         1
Name: cease_date, dtype: int64

In [59]:
dete_resignations['cease_date'] = dete_resignations['cease_date'].str.split('/').str[-1]
dete_resignations['cease_date'] = dete_resignations['cease_date'].astype("float")


In [60]:
dete_resignations['cease_date'].value_counts().sort_values()


2006.0      1
2010.0      2
2014.0     22
2012.0    129
2013.0    146
Name: cease_date, dtype: int64

In [61]:
dete_resignations['dete_start_date'].value_counts().sort_values()

1963.0     1
1971.0     1
1972.0     1
1984.0     1
1977.0     1
1987.0     1
1975.0     1
1973.0     1
1982.0     1
1974.0     2
1983.0     2
1976.0     2
1986.0     3
1985.0     3
2001.0     3
1995.0     4
1988.0     4
1989.0     4
1991.0     4
1997.0     5
1980.0     5
1993.0     5
1990.0     5
1994.0     6
2003.0     6
1998.0     6
1992.0     6
2002.0     6
1996.0     6
1999.0     8
2000.0     9
2013.0    10
2009.0    13
2006.0    13
2004.0    14
2005.0    15
2010.0    17
2012.0    21
2007.0    21
2008.0    22
2011.0    24
Name: dete_start_date, dtype: int64

In [62]:
tafe_resignations['cease_date'].value_counts()


2011.0    116
2012.0     94
2010.0     68
2013.0     55
2009.0      2
Name: cease_date, dtype: int64

The span of years for each dataframe is very different and cannto be aligned. Overall the data looks fine.

In [63]:
dete_resignations['institute_service'] = dete_resignations['cease_date'] - dete_resignations['dete_start_date']

In [64]:
dete_resignations['institute_service'].head()

3      7.0
5     18.0
8      3.0
9     15.0
11     3.0
Name: institute_service, dtype: float64

subtracting start date from the cease date gives us the length of service or duration of service for the employee

In [65]:
tafe_resignations['Contributing Factors. Dissatisfaction'].value_counts()
tafe_resignations['Contributing Factors. Job Dissatisfaction'].value_counts()

-                      270
Job Dissatisfaction     62
Name: Contributing Factors. Job Dissatisfaction, dtype: int64

In [66]:
def update_vals(val):
    if pd.isnull(val):
        return np.nan
    elif val == '-':
        return False
    else:
        return True
    
tafe_resignations['dissatisfied'] = tafe_resignations[['Contributing Factors. Dissatisfaction', 'Contributing Factors. Job Dissatisfaction']].applymap(update_vals).any(1, skipna=False)
tafe_resignations_up = tafe_resignations.copy()

dete_resignations['dissatisfied'] = dete_resignations[['job_dissatisfaction','dissatisfaction_with_the_department','physical_work_environment','lack_of_recognition','lack_of_job_security','work_location','employment_conditions','work_life_balance','workload']].any(1, skipna=False)
dete_resignations_up = dete_resignations.copy()
dete_resignations_up['dissatisfied'].value_counts(dropna=False)

    


False    162
True     149
Name: dissatisfied, dtype: int64

In [67]:
dete_resignations_up['institute'] = 'DETE'
tafe_resignations_up['institute'] = 'TAFE'

In [68]:
combined = pd.concat([dete_resignations_up, tafe_resignations_up],ignore_index=True)


In [69]:
combined.notnull().sum().sort_values()

torres_strait                                                                                                        0
south_sea                                                                                                            3
aboriginal                                                                                                           7
disability                                                                                                           8
nesb                                                                                                                 9
business_unit                                                                                                       32
Main Factor. Which of these was the main factor for leaving?                                                        96
classification                                                                                                     161
InductionInfo. Topic:Did you undertake a Corpora

In [70]:
combined_updated = combined.dropna(thresh = 500, axis =1).copy()



In [71]:
combined_updated['institute_service'].value_counts(dropna = False)



NaN                   88
Less than 1 year      73
1-2                   64
3-4                   63
5-6                   33
11-20                 26
5.0                   23
1.0                   22
7-10                  21
3.0                   20
0.0                   20
6.0                   17
4.0                   16
2.0                   14
9.0                   14
7.0                   13
More than 20 years    10
8.0                    8
13.0                   8
15.0                   7
20.0                   7
10.0                   6
12.0                   6
14.0                   6
17.0                   6
22.0                   6
18.0                   5
16.0                   5
24.0                   4
11.0                   4
23.0                   4
21.0                   3
32.0                   3
19.0                   3
39.0                   3
26.0                   2
28.0                   2
30.0                   2
25.0                   2
36.0                   2


In [72]:
# Extract the years of service and convert the type to float
combined_updated['institute_service_up'] = combined_updated['institute_service'].astype('str').str.extract(r'(\d+)')
combined_updated['institute_service_up'] = combined_updated['institute_service_up'].astype('float')

# Check the years extracted are correct
combined_updated['institute_service_up'].value_counts()

/dataquest/system/env/python3/lib/python3.4/site-packages/ipykernel/__main__.py:2: FutureWarning: currently extract(expand=None) means expand=False (return Index/Series/DataFrame) but in a future version of pandas this will be changed to expand=True (return DataFrame)
  from ipykernel import kernelapp as app


1.0     159
3.0      83
5.0      56
7.0      34
11.0     30
0.0      20
20.0     17
6.0      17
4.0      16
9.0      14
2.0      14
13.0      8
8.0       8
15.0      7
17.0      6
10.0      6
12.0      6
14.0      6
22.0      6
16.0      5
18.0      5
24.0      4
23.0      4
39.0      3
19.0      3
21.0      3
32.0      3
28.0      2
36.0      2
25.0      2
30.0      2
26.0      2
29.0      1
38.0      1
42.0      1
27.0      1
41.0      1
35.0      1
49.0      1
34.0      1
33.0      1
31.0      1
Name: institute_service_up, dtype: int64

In [73]:

def career_stage(val):
    if pd.isnull(val):
        return np.nan
    elif val < 3:
        return "New"
    elif 3 <= val <7:
        return "Experienced"
    elif 7 <= val <11:
        return "Established"
    else:
        return "Veteran"

combined_updated['service_cat'] = combined_updated['institute_service_up'].apply(career_stage)
        

In [86]:
combined_updated['dissatisfied'] = combined_updated['dissatisfied'].value_counts(dropna = False)

In [87]:
combined_updated['dissatisfied'] = combined_updated['dissatisfied'].fillna(False)


In [88]:
mean_dissatisfied= combined_updated.pivot_table(index='service_cat', values='dissatisfied')

print(mean_dissatisfied)
%matplotlib inline
mean_dissatisfied.plot(kind ='bar')

DataError: No numeric types to aggregate